README.md

```diff
@@ -629,6 +629,274 @@ Can you try updating the rendering function to also show `entry.content` and
+  Loaded(data)})`)
+- When the input value changes, we will go back to step `Idle`:
+```reasonml
+  onChange={event => {
+    setState(_ =>
+      {username: event->React.Event.Form.target##value, step: Idle}
+    )
+  }}
+```
+- We validate the username on `Enter` to ensure only valid requests are sent. If
+  the name is valid, we can just call `fetchFeed`, but if it's invalid, we will
+  set the step to `InvalidUsername`:
+```reasonml
+  onKeyDown={event => {
+    let enterKey = 13;
+    if (React.Event.Keyboard.keyCode(event) == enterKey) {
+      switch (Username.make(state.username)) {
+      | Ok(username) => fetchFeed(username)
+      | Error () =>
+        setState(state => {...state, step: InvalidUsername})
+      };
+    };
+  }}
+```
+
+To continue, we have to adapt the effect to check the username is valid:
+
+```reasonml
+  React.useEffect0(() => {
+    switch (Username.make(state.username)) {
+    | Ok(username) => fetchFeed(username)
+    | Error () =>
+      Js.Exn.raiseError(
+        "Initial username passed to React.useState is invalid",
+      )
+    };
+    None;
+  });
+```
+
+Finally, we can modify the `switch` in the JSX code to use the new state, so
+instead of checking `data` we check for `state.step`:
+
+```reasonml
+  {switch (state.step) {
+    | InvalidUsername => <div> {React.string("Invalid username")} </div>
+    | Idle =>
+      <div>
+        {React.string(
+          "Press the \"Enter\" key to confirm the username selection.",
+        )}
+      </div>
+    | Loading => <div> {React.string("Loading...")} </div>
+    | Loaded(Error(msg)) => <div> {React.string(msg)} </div>
+    | Loaded(Ok(feed)) => ...
+  }}
+```
+
+### 9.2: Handle edge cases and errors
+
+To improve the user experience, we'll add messages for the following cases:
+- when there are errors returned by the server
+- or when a valid user has an empty feed
+
+For this, we will have to modify `App` component in a couple of places.
+
+The `fetchFeed` function should check for the response status using
+`Fetch.Response.status`. If the status is 200, then proceed as before, but if
+it's something else, it should call `setData(_ => Loaded(Error("Error: Received
+status " ++ string_of_int(status))))`.
+
+The code should roughly look like this:
+
+```reasonml
+Fetch.fetch(...)
+|> P.then_(response => {
+      let status = Fetch.Response.status(response);
+      if (status === 200) {
+        /* If status is OK, proceed to parse the response */
+        response
+        |> Fetch.Response.text
+        |> P.then_(...);
+      } else {
+        /* Handle non-200 status */
+        setState(state =>
+          {
+            ...state,
+            step:
+              Loaded(
+                Error(
+                  "Error: Received status " ++ string_of_int(status),
+                ),
+              ),
+          }
+        )
+        |> P.resolve;
+      };
+    })
+|> ignore;
```

Comment:
I think this template helps a lot but I forgot to place a `P.resolve`. Incidentally, one thing I noticed is the error message you see when you use module alias was a little easier to understand than the one you get if you use local open.

Module alias:

```
File "src/App.re", line 69, characters 18-68:
69 |                   setState(state => {...state, step: Loaded(data)});
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This expression has type unit but an expression was expected of type
         'a P.t
```

Local open:

```
File "src/App.re", line 70, characters 18-68:
70 |                   setState(state => {...state, step: Loaded(data)});
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This expression has type unit but an expression was expected of type
         'a t
```